### DISCUSSION TITLE
Changesets - releases and changelogs

### DISCUSSION THREAD

**Discussion by eliashussary** [directus/directus#5698] [Feature Requests]
@benhaynes @rijkvanzanten 

I know there are a few discussions on this matter, so I know I'm not the only one looking for these :)

I'm aware we have milestones. I guess my only gripe with it is that it lists a bunch of issues. There isn't much I can gather from it without digging into each of them.

I propose that this project adopts https://github.com/atlassian/changesets or something similar to this.

I understand there was trouble in the past with conventional-commits, surely an issue for any open source project is this type of adherence to rules at scale.

The nice thing with changesets is that it becomes part of the PR process, and doesn't need to be done by the contributor. The maintainers are free to do it themselves if that's the direction.  It becomes super clear what effect the change has on versioning and it is explicit, and every change has clear meaningful readable messages. - Multiple PR's can also be grouped as a changeset, and the changeset message can be modified to be more concise. 

Changesets CLI and GH Action helps automate changelog generation, versioning, and publishing.

While this may not be the chosen tool long term, this may provide a glimpse into alternatives to things like conventional commits.

Our team has migrated from conventional-commits to changesets and have had a much better experience with managing this aspect of our operations.

I understand we're in a state where we're technically still pre-9.0 and change is frequent... but i feel like every couple of days i'm updating and have no idea what's going on at a quick glance :). Naturally as maintainers you're close to the problems and releases, but i'm failing to keep up :( - i'd love it to be easier to consume this info quickly.

**Comment by rijkvanzanten** [comment]
Haven't heard of that tool! Worth taking a look.

The main issue with a commit based strategy is enforcement. I can't reliably expect everybody (including myself frankly) to adhere to those rules.

I do agree that having better changelogs is a definite nice to have. I think that having a way to automatically pull the issues from a milestone and format / display them by type in the release notes could be a good middle ground in case that tooling doesn't work out

**Reply by benhaynes** [reply]
I agree that we need to find a good balance here, as the release notes need to be more verbose/complete. Perhaps there's an automated way to include issue titles/details in the release message? I know @rijkvanzanten is saying it's hard to get good input in those issues, but I can't think of any other way to avoid this being a completely manual process.

**Comment by rijkvanzanten** [comment]
Changesets looks promising! I am a little worried by the lack of documentation, but seeing Lerna is.. worse in that regard and seemingly no longer maintained, I think it's worth trying it out 🙂

**Reply by rijkvanzanten** [reply]
We'll have to wait until the v9.0.0 release though, as it doesn't seem like we'll be able to use changesets with a custom version tag 🤔

**Reply by eliashussary** [reply]
That's a good point, not that I'm aware of. Looks like they have a `prerelease` concept but at the cost of more complexity.

https://github.com/atlassian/changesets/blob/main/docs/prereleases.md  ... doesn't look like much fun :(

I suppose the question is, what's your intended strategy for releases in future? i.e. do you plan on having `rc`'s or just have a rolling `nightly` or `next` tag, etc?

**Reply by rijkvanzanten** [reply]
Once v9.0.0 is released, we'll (obviously) have "regular" releases. It would be nice if we could do an automated nightly push, but that's not a hard requirement. The only times we'll use a `next` tag would be for big changes that need community review before release 🤔

**Reply by eliashussary** [reply]
What if for simplicity sake nightly was just published as `nightly` and if you needed a snapshot version you can simply use something more along the lines of datever `nightly-2021.05.17.01` -- or try and somehow incorporate the next potential version out of changset without actually publishing?

**Reply by rijkvanzanten** [reply]
Yeah that makes sense. We also don't necessarily have to publish changelogs/tags/releases for nightly releases, as they're mostly for testing / debugging purposes 🤔 That's a problem we can solve for later! (We'd have to make a new nightly branch for the versioning changes not to intervene with `main` anyways

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
